js-packages/@prestojs/rest/src/useAsyncValue.ts

```diff
@@ -0,0 +1,249 @@
+     */
+    error: null | Error;
+    /**
+     * True while `resolve` call is in progress.
+     */
+    isLoading: boolean;
+    /**
+     * The resolved value
+     */
+    value: T | null;
+    /**
+     * A function to manually trigger the action. If `options.trigger` is `MANUAL`
+     * calling this function is the only way to trigger the action.
+     *
+     * This function will return a promise that resolves/rejects to same value
+     * returned by `execute`. If `accumulatePages` is set the value returned is
+     * the accumulated value.
+     */
+    run: (...args) => Promise<any>;
+    /**
+     * When called will unset result and error to null. Will not immediately trigger
+     * a call to the action but subsequent changes to query or paginator will according
+     * to the value of `trigger`.
+     *
+     * If value is found in `existingValues` then it will still be returned even if
+     * you call `reset` - `reset` only clears values that are returned from `resolve`.
+     */
+    reset: () => void;
+};
+
+/**
+ * Resolve a value from an id using an async function.
+ *
+ * For the specified `id` the `resolve` function will be called and should
+ * return the value for the specified id.
+ *
+ * If `existingValues` is provided this will be consulted first before calling
+ * `resolve`. This is useful when you have a list of values that may or
+ * may not have the data you care about. If it's there then the value will
+ * be returned immediately.
+ *
+ * For multiple values see documentation below.
+ *
+ * @extract-docs
+ */
+export default function useAsyncValue<T, U extends Id>(
+    props: UseAsyncValuePropsSingle<T, U>
+): UseAsyncValueReturn<T>;
+/**
+ * Resolve values from an array of ids using an async function.
+ *
+ * For the specified array of `ids` the `resolve` function will be called
+ * and should return an array of the same size with each entry being the resolved
+ * value for the corresponding id.
+ *
+ * If `existingValues` is provided this will be consulted first before calling
+ * `resolve`. This is useful when you have a list of values that may or
+ * may not have the data you care about. If it's there then the values will
+ * be returned immediately. Note that if any of the ids are missing from
+ * `existingValues` then it will be ignored and a call to `resolve` will be made
+ * requesting values for all `ids`.
+ */
+export default function useAsyncValue<T, U extends Id>(
+    props: UseAsyncValuePropsMulti<T, U>
+): UseAsyncValueReturn<T[]>;
+export default function useAsyncValue<T, U extends Id>(
+    props: CommonProps<T, U> & {
+        id?: U | null;
+        ids?: U[] | null;
+        resolve: (idOrIds: U | U[]) => Promise<T[] | T>;
+    }
+): UseAsyncValueReturn<T | T[]> {
+    const { id, ids, getId, resolve, trigger } = props;
+    if (id && ids) {
+        throw new Error("Only one of 'id' and 'ids' should be provided");
+    }
+    const idRef = useRef(id || ids);
```

Comment:
the reason is so you can read the value of the ref in the memoized `execute` function below. eg. whenever `execute` is called it uses the last value set. Using this we can support passing inline functions to `useAsyncValue`... without it you'd be forced to memoize it yourself all the time. The downside of it is if the _only_ thing that changes is the function then it won't trigger a change but for the use case here I think this is acceptable (ie. the only thing that will recalculate the value is a change in `id`/`ids` or a call to `run` directly. This is similar to `useSelector` in redux - it uses the same approach from memory.

Perhaps worth some more thought / discussion... in particular I do that here but I don't do it in `useAsync`. Perhaps `useAsync` should do it too but seems riskier to me given it's a more general purpose hook

Reply:
I’ve wasted hours on before on react components that sometimes don’t react to props changes as you would expect.

Do you have an example of what the dev/code cost would be if we simply respond to the execute function changing without trying to add any conveniences?

Reply:
Sure, if we make it a dep then someone writes this:

```ts
const retrieveDeps = asyncChoices.useRetrieveDeps({});
const selected = useAsyncValue({
     ids: value,
     resolve: props => asyncChoices.retrieve(props, retrieveDeps),
});
```

they get runaway effect/infinite loop

can't just pass deps because object will change, need to make sure identity is constant

```ts
// Can't just use this any more
const retrieveDeps = asyncChoices.useRetrieveDeps({});
// We know we need pagiantor so pull that out. Making this entirely generic
// (eg. for other deps that might be in asyncChoices) is more work 
const { paginator } = retrieveDeps;
const resolve = useCallback(
    props => asyncChoices.retrieve(props, { paginator } )),
    [paginator]
);
const selected = useAsyncValue({
     ids: value,
    resolve,
});
```